### ISSUE TITLE
few questions about usage

### ISSUE THREAD

**Issue by jradom** [moshensky/pdf-visual-diff#17]
Hello,

Please, let me know if your library can have the following parameters:
threshold
mask areas to exclude comparisons
it's not clear to me how to use it with Typescript
Please, clarify

Thanks

jeff

**Comment by jradom** [comment]
allow include/exclude whole pages?

**Comment by jradom** [comment]
I'm very new to Typescript and please clarify what did you mean by
"If you are using Typescript add import('pdf-visual-diff/lib/toMatchPdfSnapshot') to your typings.".  How/where to put it in the code

**Comment by jradom** [comment]
It'll be really great if I can get a simple example code of how to compare 2 page PDF files with an exclusion rectangle, only first page, and diff file.  Also, with a minimum include files etc.

Thanks

  Jeff

**Comment by moshensky** [comment]
Hello @jradom,

- threshold: all heavy lifting underneath is done by [GraphicsMagick](http://www.graphicsmagick.org/). Right now you can pass a `tolerance` value via `compareImageOpts` argument. GM's compare function is used to compare to images, which returns `equality`. Comparing is considered successful when returned `equality` is `<=` than passed in `tolerance` value, which by default is 0 (i.e. images must match exactly pixel by pixel).
- mask areas to exclude comparisons: right now rectangles are supported. For more info check this issue:  #15.
- allow include/exclude whole pages: this is something that is not supported by this lib, although it should be trivial to add, but it is some work that will take some time. Meanwhile there are libs out there that could be used to split pdf into pages and then feed them here. 
- "If you are using Typescript add import('pdf-visual-diff/lib/toMatchPdfSnapshot') to your typings.": this is useful in conjunction with jest. For example in a project where I use this lib, next to `src` dir I have `typings` where I have `index.d.ts` file. In that file I have a line with `import('pdf-visual-diff/lib/toMatchPdfSnapshot')` entry that is used for type inference by tsc. Of course custom jest matcher needs to be configured via jest configuration files. Also in my `tsconfig.json` file there is an entry that includes those typings `"files": ["./typings/index.d.ts"]`.
- simple example code of how to compare 2 page PDF: I will do my best to commit a sample project today. 

@jradom could you provide me with a sample 2 page pdf and describe which part you'd like to be masked?

**Comment by jradom** [comment]
Can you please tel me if coordinates of exclusion rectangles are in pixels or millimeters?  I still didn't get if let's say I have PDF files consistin of 10 pages but only want to compare only first 3 or just pages of my choice?  Is it currently possible?   I also interested if I can somehow call comparison from other code for example Cypress if you heard about it.  It also supports Mocha and Chai

Thanks in advance

Jeff

**Comment by jradom** [comment]
I mean how can I call your functions from code like thsi?

describe('PDF reports comparison', () => {
  it('should compare PDF files', () => {

    /*  beforeEach(() => {
        cy.wait(2000)
      })*/
    cy.pause();
    cy.visit('baselineFinalReport.pdf');
    cy.log('VVVVVVVVVVVVVVVVVVVVVV');
/// Call for pdf-visual diff 
/// compare baseline PDF with actual PDF here
  });
})

What should include in order to make it work?

**Comment by jradom** [comment]
"target": "es6",
Will your code work with the above version?

**Comment by jradom** [comment]
we plan to use es6 for now

**Comment by jradom** [comment]
I'm getting these results when running your sample in IntelliJ
Please, advise

Thanks

Jeff

![image](https://user-images.githubusercontent.com/20550461/98016061-60c77280-1db2-11eb-9301-4b1def58cd3b.png)

**Comment by moshensky** [comment]
> Can you please tel me if coordinates of exclusion rectangles are in pixels or millimeters? I still didn't get if let's say I have PDF files consistin of 10 pages but only want to compare only first 3 or just pages of my choice? Is it currently possible? I also interested if I can somehow call comparison from other code for example Cypress if you heard about it. It also supports Mocha and Chai

- Coordinates are in pixels after converting pdf with density(300, 300).
- Right now you can't define which pages to compare. All pages are converted to one big image that is being compared. In order to compare some of the pages you'd have to split it by other means.
- Testing frameworks are irrelevant, because this lib exports a function `comparePdfToSnapshot` that returns a Promise<boolean>, which you can incorporate anywhere.
 



> I mean how can I call your functions from code like this?
> 
> describe('PDF reports comparison', () => {
> it('should compare PDF files', () => {
> 
> ```
> /*  beforeEach(() => {
>     cy.wait(2000)
>   })*/
> cy.pause();
> cy.visit('baselineFinalReport.pdf');
> cy.log('VVVVVVVVVVVVVVVVVVVVVV');
> ```
> 
> /// Call for pdf-visual diff
> /// compare baseline PDF with actual PDF here
> });
> })
> 
> What should include in order to make it work? Just minimum code for your package which can be used in my own code

I failed to find time to make a sample. It is on my asap list of tasks. I will do it tomorrow morning.

**Comment by moshensky** [comment]
> I'm getting these results when running your sample in IntelliJ
> Please, advise
> 
> Thanks
> 
> Jeff
> 
> ![image](https://user-images.githubusercontent.com/20550461/98016061-60c77280-1db2-11eb-9301-4b1def58cd3b.png)

Could you provide as text full output please. Please tell me what version of node and graphicsmagick you are using. Could you run all test from the terminal (npm run test) and provide its output too?

**Comment by jradom** [comment]
I've tried this code based on one of your samples
  it('Roche should create new snapshot, when one does not exists', () => {
    const snapshotName = 'draftReport'
    const snapshotPath = join(__dirname, snapshotsDirName, snapshotName + '.png')
    if (existsSync(snapshotPath)) {
      unlinkSync(snapshotPath)
    }

    return comparePdfToSnapshot(baselineDraftReportPath, __dirname, snapshotName).then((x) => {
      expect(x).to.be.true
      expect(existsSync(snapshotPath)).to.be.true
      unlinkSync(snapshotPath)
    })
  })

And I can't see/find a snapshot file being created.  

Please, advise

Thanks

jeff

**Comment by jradom** [comment]
Also, please suggest how the above code can be simplified?

**Comment by jradom** [comment]
Also is it a tool or maybe I can use Acrobat Pro to create baseline images out of PDF files?

**Comment by jradom** [comment]
can I, for example, use these conversions which is part Acrobat Pro?

![image](https://user-images.githubusercontent.com/20550461/98084451-a9be0c00-1e30-11eb-976c-5634c5da1457.png)
And if I can, I might need to set the resolution, size, etc of a target image

**Comment by moshensky** [comment]
@jradom please find here a bare-bones sample project: https://github.com/moshensky/pdf-visual-diff/tree/master/examples



> Also, please suggest how the above code can be simplified?

Please provide me with a link to repository where I could take a look at what you have. It would be best to have there only the problematic bits.



> Also is it a tool or maybe I can use Acrobat Pro to create baseline images out of PDF files?

No. If there are no snapshots for a pdf to be compared, then on the first run they are created automatically. After that they are used to compare. Please checkout the bare-bones sample and play with it.


When I find a bit more free time I will create a few more complex examples using TS and exercise various options that could be used (masking regions, treshold, etc).

**Comment by jradom** [comment]
I now want to try using your toll just to compare screenshots. 

I had used this below sample code
import { compareImages, mkDiffPath } from './compare-images'
import { expect } from 'chai'
import { join } from 'path'
import { existsSync } from 'fs'

describe('mkDiffPath()', () => {
  it('should mk path with extension', () =>
    expect(mkDiffPath('some-path.ext')).to.equal('some-path.diff.ext'))

  it('should mk path with extension when starts with .', () =>
    expect(mkDiffPath('./some-path.ext')).to.equal('./some-path.diff.ext'))

  it('should handle empty', () => expect(mkDiffPath('')).to.equal('.diff'))

  it('should mk path without extension', () =>
    expect(mkDiffPath('some-path')).to.equal('some-path.diff'))
})

///const expectedImageName = 'sample-image-expected.png'
const expectedImageName = 'draftReport-expected.png'
const sampleImage = 'sample-image.png'
const sampleImage2 = 'sample-image-2.png'
const sampleImage3 = 'draftReport.png'
///const sampleImage3 = 'sample-image-3.png'
const testDataDir = join(__dirname, './test-data')
const expectedImagePath = join(testDataDir, expectedImageName)
const imagePath = join(testDataDir, sampleImage)
const image2Path = join(testDataDir, sampleImage2)
const image3Path = join(testDataDir, sampleImage3)

describe('compareImages()', () => {
  it('should succeed comparing', () =>
    compareImages(expectedImagePath, imagePath).then((x) => {
      expect(x).to.be.true
      expect(existsSync(mkDiffPath(imagePath))).to.eq(false, 'should not generate diff output')
    }))

  it('should fail comparing and output diff', () =>
    compareImages(expectedImagePath, image2Path).then((x) => {
      expect(x).to.be.false
      expect(existsSync(mkDiffPath(image2Path))).to.be.true
    }))

  it('should fail comparing without outputting diff', () =>
    compareImages(expectedImagePath, image3Path, { writeDiff: true, highlightColor: 'Red', highlightStyle: 'Tint'}).then((x) => {
      expect(x).to.be.false
      expect(existsSync(mkDiffPath(image3Path))).to.be.false
    }))
})


How/where can I set parameters like threshold, multiple masks, etc.  Also, is it possible to draw red rectangles around mismatched areas?

Please, let me know

Thanks

Jeff

**Comment by moshensky** [comment]
Here are the options that can be passed: https://github.com/moshensky/pdf-visual-diff/blob/master/src/compare-images.ts#L35
You need to set `maskRegions` in order to have rectangle masks, something along the lines:
```js
compareImages(expectedImagePath, image3Path, { 
  writeDiff: true, 
  highlightColor: 'Red', 
  highlightStyle: 'Tint',
  maskRegions: [
    {
          type: 'rectangle-mask',
          x: 120,
          y: 150,
          width: 280,
          height: 200,
          color: 'Blue',
        }
  ],
})
```
Like it is done here https://github.com/moshensky/pdf-visual-diff/blob/master/src/compare-pdf-to-snapshot.spec.ts#L62. Please note that `comparePdfToSnapshot` and `compareImages` accept same type for options.

It is not possible to draw rectangles around mismatched areas at the moment. You can play with `highlightColor` and `highlightStyle` that are used in the generated diffs when a mismatch is found.

**Comment by jradom** [comment]
can your toll crop images to be compared?  If yes, how can I do that?

Thanks

Jeff

**Comment by moshensky** [comment]
Do you mean library by toll? 
I'm afraid I can't fully grasp your question. Could you rephrase it bit? 
The lib is created solely to take image snapshots of pdfs and later compare does snapshots to same pdfs.
I use this lib in my tests in a few projects to compare generated pdf reports in order to catch any breaking changes due to code changes.

**Comment by jradom** [comment]
toll was a typo :) I meant "tool".  Like I said because of dealing with pages to skip or choose to compare, I plan to use your tool  to compare screenshots for now

**Comment by jradom** [comment]
Oh, let's say I have a 4 page PDF report file and want to compare only the first page for example and also on the first page want to exclude the footer and header.  How can I do this with the masks?   Like 2 masks for footer and header and 3 masks to cover pages 2-4?  Please, help

Thanks

Jeff

**Comment by moshensky** [comment]
> Oh, let's say I have a 4 page PDF report file and want to compare only the first page for example and also on the first page want to exclude the footer and header. How can I do this with the masks? Like 2 masks for footer and header and 3 masks to cover pages 2-4? Please, help
> 
> Thanks
> 
> Jeff
                           
That should work. Pdfs are converted to a single page, i.e. a multi page pdf will be converted to a single image. Which means that you have to provide a minimum of 2 masks. First needs to cover header and second starting from footer's yPos with height to the bottom of the image.
To find right masks coordinate you could:
- compare pdf for the first time, which will just create a new snapshot
- open created snapshot in any image viewer where you can see coordinates in pixels
- figure out from there what xs,ys, widths and heights you need to use for your masks
- update function from step 1 to use masks with appropriate coordinates

**Comment by jradom** [comment]
Hi Nikita,

Can you please make for me a small project as you had suggested with 2
simple with few options PDF comparison (positive and negative) and 2 Image
comparison?  Also, how do you suggest to update baseline images and PDF
reports?  BTW, I can speak Russian because was born and raised in
Leningrad.  I had left USSR about 30 years ago and live in California now

Thanks a lot in advance

Jeff

**Comment by jradom** [comment]
Hi Nikita,

I also a  bit confused and will explain you how I compare 2 PDF reports or images with existing  ode in Java. And want to maintain similar flow with yours. I create a baseline pdf or png file and keep those until I need to change those. The in case of PDF I generate a new pdf with the same name and compare baseline with actual pdf report. In other words I compare existing PDF with new ones. How can I do it with yours?  It looks to me that you compare baseline image of pdf with actual pdf, correct. Please clarify 

Thanks

Jeff

**Comment by moshensky** [comment]
Hi @jradom,

I'm from Leningrad as well, but live in Bulgaria since 30 years :). If you wish to continue in Russian, let's use some other channels to communicate. 

In the next few days I can create another sample project, where there are going to be one passing and one failing cases.

The lib indeed compares images of pdfs. First pdfs are converted to images and then being compared to a `baseline` as you call it. 
For your particular case I'd advice to create a baseline pdf. Put it in a well known place and add some code like in the example that I made for you a few comments above. Run the code for the first time. This should generate a snapshot in form of an image of the pdf. All consequence runs will assert whether the pdf from the `well known place` has been changed since last run. When you generate new report and want to compare it, you would have to replace the file in that `well known place` with the new one and run the program again. If differences would be found you will find next to the snpashot two new files: one with a `new` suffix and one with a `diff` prefix. The diff will show where are the differences and the new one is... well the new image made from the pdf. If you would want to keep the changes you delete the old snpashot, the diff and rename the file with `new` prefix by removing the prefix.

**Comment by jradom** [comment]
What a coincidence.   I left in 1989.  So,  I'd prefer to have
something like this
[image: image.png]

I also prefer if possible to have a sample in Typescript.  If just image
comparison in addition to having this added to PDF comparisons.   it'll be
really great.  I'll ned eventually to include/call this code into
Typescript/cypress which we are.  I have a problem writing in Russian not
because I forgot it but the Russian keyboard, etc.  I'm curious how you got
to Bulgaria?  My friends have some property there.  I wish I can go
someday there but too many hopes from san Francisco.  BTW, I did use the
PDF Compare Java library almost 2 years ago.  But go away from Java to
Typescript/Cypress to UI testing.

Thanks a lot in advance

Jeff

**Comment by moshensky** [comment]
Hi @jradom, 

sorry for late reply, do you still have a need for TS sample?

**Comment by jradom** [comment]
Yes

**Comment by jradom** [comment]
Hi Nikita,

I'm thinking to use your tool for both PNG and PDF file comparisons.  So,
I'll greatly appreciate if you provide me with some stuff

Thanks a lot in advance

Jeff

**Comment by moshensky** [comment]
Hi Jeff,

I'll do my best to scaffold a sample project this evening or tomorrow morning.

**Comment by jradom** [comment]
sounds like a great plan.  Something very simple which I can reuse with
some changes in other projects.  Just two positive/negative PDF or/and PNG
comparisons

Thanks a lot in advance

Jeff

**Comment by moshensky** [comment]
Hi Jeff,

I have added a sample project in typescript. You can find it here https://github.com/moshensky/pdf-visual-diff/tree/master/examples/ts-sample.
I will try to add more sample usages in the upcoming days.

**Comment by moshensky** [comment]
Closing. Please feel free to reopen with new questions.
